### PR TITLE
Use dataclass instead of pytools.Record

### PR DESCRIPTION
**Pull request by alexfikl**
The change in `reduction.py` has some actual logic changes (made `_ReductionInfo` frozen so had to move things around), but the rest should be cleanup.

Most of these classes seemed internal, so I assume they don't need some compatibility with `Record` (?).

### CODE REVIEW

**Comment by inducer** [comment]
Thanks! I'm open to a mutually dependent change in boxtree to help us get rid of `Record`.

**Comment by alexfikl** [comment]
> Thanks! I'm open to a mutually dependent change in boxtree to help us get rid of `Record`.

I think I just missed a member in `BuiltList`, so hopefully the latest version will fix `boxtree`.

File: pyopencl/algorithm.py [hunk 1]
```diff
@@ -29,13 +29,19 @@
 OTHER DEALINGS IN THE SOFTWARE.
 """
 
+from dataclasses import dataclass
```
Comment (inducer) [inline pyopencl/algorithm.py:32]: The compat target for pyopencl is nominally still Py3.6, so this needs a conditional dependency in `setup.py`.

Comment (alexfikl) [reply pyopencl/algorithm.py:32]: Right! Added in [f9e31f4](https://github.com/inducer/pyopencl/pull/571/commits/f9e31f41439c9a7fdfc0ae8476a05f4d8388e125).

File: pyopencl/array.py [hunk 2]
```diff
@@ -2304,20 +2305,19 @@ def arange(queue, *args, **kwargs):
...
-    class Info(Record):
-        pass
+    @dataclass
+    class _ArangeInfo:
+        start: Optional[int] = None
+        stop: Optional[int] = None
+        step: Optional[int] = None
+        dtype: Optional["np.dtype"] = None
+        allocator: Optional[Any] = None
+        wait_for: List[cl.Event] = field(default_factory=list)
```
Comment (inducer) [inline pyopencl/array.py]: I think this might be better as a class-local.

**Review by inducer — APPROVED** [review]
Thanks! Just one request, otherwise this LGTM!

File: pyopencl/array.py (hunk 2 above)
Comment (alexfikl) [reply pyopencl/array.py]: Function-local? Just `wait_for`, right?

Comment (inducer) [reply pyopencl/array.py]: What I mean is that I would prefer to not repeatedly incur the cost for code generation in `@dataclass`.

Comment (alexfikl) [reply pyopencl/array.py]: Ah, completely misunderstood that, sorry!

Moved it out of the function in https://github.com/inducer/pyopencl/compare/f9e31f41439c9a7fdfc0ae8476a05f4d8388e125..b0a975c4433fff5659fe55d8da3f117788bf199b.

**Comment by inducer** [comment]
Thx!

**Pull request merged**